core/common/src/main/java/alluxio/security/authentication/AuthenticatedUserInjector.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.security.authentication;
+
+import alluxio.exception.status.UnauthenticatedException;
+
+import io.grpc.ForwardingServerCallListener;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+
+import java.util.UUID;
+
+/**
+ * Server side interceptor for injecting authenticated user into TLS. This interceptor requires
+ * {@link ChannelIdInjector} to have been injected the channel id from which the particular RPC is
+ * being made.
+ */
+public class AuthenticatedUserInjector implements ServerInterceptor {
+
+  private AuthenticationServer mAuthenticationServer;
+
+  /**
+   * Creates {@link AuthenticationServer} with given authentication server.
+   * @param authenticationServer the authentication server
+   */
+  public AuthenticatedUserInjector(AuthenticationServer authenticationServer) {
+    mAuthenticationServer = authenticationServer;
+  }
+
+  @Override
+  public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+      Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+    return new ForwardingServerCallListener.SimpleForwardingServerCallListener<ReqT>(
+        next.startCall(call, headers)) {
+      /**
+       * onHalfClose is called on the same thread that calls the service handler.
+       */
+      @Override
+      public void onHalfClose() {
+        // Try to fetch channel Id from the metadata.
+        UUID channelId = headers.get(ChannelIdInjector.S_CLIENT_ID_KEY);
+        if (channelId != null) {
+          try {
+            // Fetch authenticated username for this channel and set it.
+            String userName = mAuthenticationServer.getUserNameForChannel(channelId);
+            if (userName != null) {
+              AuthenticatedClientUser.set(userName);
+            } else {
+              AuthenticatedClientUser.set("");
```

Comment:
when do we AuthenticatedClientUser.set to empty vs AuthenticatedClientUser.remove?

Reply:
Good question.This comes down to another question: "Do we want to allow unauthenticated access to a service that requires authentication". 

For example, in thrift, during client startup we open an unauthenticated transport to the MetaMaster to load the configuration. However rest of the clients to it are authenticated.

Reply:
If a service requires authentication, we should not allow unauthenticated access. The services should be split up in a way that either the entire service is authenticated or not at all (do not allow no-auth for some calls only)

Reply:
Made these changes to address this comment:
* Split metamaster client service into client and config-client services. 
* Disable authentication for config-client service.
* Fail RPCs in interceptor if an authenticated service is being called with no channel-id.